docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,387 @@
+     }
+   }
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where taint may flow from.
+-  ``isSink`` - defines where taint may flow to.
+-  ``isSanitizer`` - optionally, restricts the taint flow.
+-  ``isAdditionalTaintStep`` - optionally, adds additional taint steps.
+
+Similar to global data flow, the characteristic predicate (``MyTaintTrackingConfiguration()``) defines the unique name of the configuration and the taint analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``.
+
+Predefined sources and sinks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The data flow library contains a number of predefined sources and sinks, providing a good starting point for defining data flow based security queries.
+
+-  The class ``RemoteFlowSource`` (defined in module ``codeql.ruby.dataflow.RemoteFlowSources``) represents data flow from remote network inputs. This is useful for finding security problems in networked services.
+-  The library ``Concepts`` (defined in module ``codeql.ruby.Concepts``) contains several subclasses of ``DataFlow::Node`` that are security relevant, such as ``FileSystemAccess`` and ``SqlExecution``.
+
+For global flow, it is also useful to restrict sources to instances of ``LocalSourceNode``.
+The predefined sources generally do that.
+
+Class hierarchy
+~~~~~~~~~~~~~~~
+
+-  ``DataFlow::Configuration`` - base class for custom global data flow analysis.
+-  ``DataFlow::Node`` - an element behaving as a data-flow node.
+    -  ``DataFlow::LocalSourceNode`` - a local origin of data, as a data-flow node.
+    -  ``DataFlow::ExprNode`` - an expression behaving as a data-flow node.
+    -  ``DataFlow::ParameterNode`` - a parameter data-flow node representing the value of a parameter at method/block entry.
+
+    -  ``RemoteFlowSource`` - data flow from network/remote input.
+    -  ``Concepts::SystemCommandExecution`` - a data-flow node that executes an operating system command, for instance by spawning a new process.
+    -  ``Concepts::FileSystemAccess`` - a data-flow node that performs a file system access, including reading and writing data, creating and deleting files and folders, checking and updating permissions, and so on.
+    -  ``Concepts::Path::PathNormalization`` - a data-flow node that performs path normalization. This is often needed in order to safely access paths.
+    -  ``Concepts::CodeExecution`` - a data-flow node that dynamically executes Python code.
+    -  ``Concepts::SqlExecution`` - a data-flow node that executes SQL statements.
+    -  ``Concepts::HTTP::Server::RouteSetup`` - a data-flow node that sets up a route on a server.
+    -  ``Concepts::HTTP::Server::HttpResponse`` - a data-flow node that creates an HTTP response on a server.
+
+-  ``TaintTracking::Configuration`` - base class for custom global taint tracking analysis.
+
+Examples
+~~~~~~~~
+
+This query shows a data flow configuration that uses all network input as data sources:
```

Comment:
Is it possible to expand the explanations in this section a little?

Reply:
I've pushed a commit that expands these.

Reply:
Thank you 💖